cmd/gpu_plugin/gpu_plugin.go

```diff
@@ -778,70 +782,103 @@ func main() {
-		os.Exit(1)
+	// Validate preferred allocation policy
+	if !(opts.preferredAllocationPolicy == "balanced" || opts.preferredAllocationPolicy == "packed" || opts.preferredAllocationPolicy == "none") {
+		klog.Fatal("Invalid value for preferredAllocationPolicy. Valid values: balanced, packed, none")
 	}
 
 	klog.V(1).Infof("GPU device plugin started with %s preferred allocation policy", opts.preferredAllocationPolicy)
 
-	plugin := newDevicePlugin(prefix+sysfsDrmDirectory, prefix+devfsDriDirectory, opts)
+	return opts
+}
 
+func checkWSLMode(plugin *devicePlugin) {
+	// WSL mode validation and feature handling
 	if plugin.options.wslScan {
 		klog.Info("WSL mode requested")
 
 		if plugin.options.resourceManagement {
-			klog.Error("Resource management is not supported within WSL. Please disable resource management.")
-
-			os.Exit(1)
+			klog.Fatal("Resource management is not supported within WSL. Please disable resource management.")
 		}
 
 		if plugin.options.enableMonitoring {
-			klog.Error("Monitoring is not supported within WSL. Please disable monitoring.")
-
-			os.Exit(1)
+			klog.Fatal("Monitoring is not supported within WSL. Please disable monitoring.")
 		}
 
 		if plugin.options.healthManagement {
-			klog.Error("Health management is not supported within WSL. Please disable health management.")
-
-			os.Exit(1)
+			klog.Fatal("Health management is not supported within WSL. Please disable health management.")
 		}
 	}
+}
 
+func initializeServices(plugin *devicePlugin) {
 	if plugin.options.healthManagement || plugin.options.wslScan {
 		plugin.levelzeroService = levelzeroservice.NewLevelzero(gpulevelzero.DefaultUnixSocketPath)
-
 		go plugin.levelzeroService.Run(true)
 	}
+}
 
+func handleResourceManagement(plugin *devicePlugin) {
 	if plugin.options.resourceManagement {
-		// Start labeler to export labels file for NFD.
+		// Start labeler to export labels file for NFD
 		nfdFeatureFile := path.Join(nfdFeatureDir, resourceFilename)
-
 		klog.V(2).Infof("NFD feature file location: %s", nfdFeatureFile)
 
-		// Labeler catches OS signals and calls os.Exit() after receiving any.
+		// Labeler catches OS signals and calls os.Exit() after receiving any
 		go labeler.Run(prefix+sysfsDrmDirectory, nfdFeatureFile,
 			labelerMaxInterval, plugin.scanResources, plugin.levelzeroService, func() {
 				// Exit the whole app when labeler exits
 				os.Exit(0)
 			})
 	}
+}
+
+func main() {
+	// Parse command-line options and validate them
+	opts := parseAndValidateOptions()
+
+	// Create the device plugin
+	plugin := newDevicePlugin(prefix+sysfsDrmDirectory, prefix+devfsDriDirectory, opts)
+
+	// Handle WSL mode and exit if not supported
+	checkWSLMode(plugin)
+
+	// Initialize health management or WSL scan service if needed
+	initializeServices(plugin)
+
+	// Handle resource management if enabled
+	handleResourceManagement(plugin)
 
+	// Start the manager to run the plugin
```

Comment:
Thanks! This makes the main much cleaner.

Reply:
Thx